Gosu/Gosu.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+// #include <Gosu/Window.h>
+#include <Gosu/Image.h>
+// #include <Gosu/Font.h>
+// #include <Gosu/TextInput.h>
+// #include <Gosu/Sample.h>
+// #include <Gosu/Song.h>
+// #include <Gosu/Channel.h>
+// #include <Gosu/Color.h>
+// #include <Gosu/GLTexInfo.h>
+
+// #include "Gosu/Buttons.h"
+
+// Graphics operations
+void Gosu_gl(*fun_ptr, &func);
+void Gosu_flush();
+Gosu_Image Gosu_render(int width, int height, *fun_ptr &func);
+Gosu_Image Gosu_record(int width, int height, *fun_ptr &func);
+void Gosu_clip_to(double x, double y, double width, double height, *fun_ptr &func);
+
+// Transformations
+void Gosu_transform();
+void Gosu_translate();
+void Gosu_scale();
+void Gosu_rotate();
+
+// Rendering
+void Gosu_draw_line();
+void Gosu_draw_rect(double x, double y, double width, double height, Gosu_Color c, int mode);
+void Gosu_draw_quad();
+void Gosu_draw_triangle();
+
+// Math functions
+double Gosu_angle(double x1, double y1, double x2, double y2);
+double Gosu_distance(double x1, double y1, double x2, double y2);
+double Gosu_angle_diff(double angle1, double angle2);
+double Gosu_offset_x(double theta, double r);
+double Gosu_offset_y(double theta, double r);
+double Gosu_random(double min, double max);
+
+// Window/Screen information
+int Gosu_available_width();
+int Gosu_available_height();
+int Gosu_screen_width();
+int Gosu_screen_height();
+
+// Button querying
+bool Gosu_button_down(int id);
+char* Gosu_button_id_to_char(int id);
+int Gosu_button_char_to_id(char* character);
+
+// Misc
+int Gosu_fps();
+char* Gosu_language();
```

Comment:
The functions returning `char*` will be interesting. Probably easiest to keep a `static thread_local std::string` in the wrapper functions so that the returned `const char*` can be converted into a Ruby string without any potential race conditions when multiple threads try to read it. (Not that Gosu itself is thread-safe right now...)

Reply:
Is this how `static thread_local std::string` should be used for the desired affect?
```C
  const char* Gosu_Image_to_blob(Gosu_Image* image)
  {
    Gosu::Image* gosu_image = reinterpret_cast<Gosu::Image*>( image );
    static thread_local std::string blob = (const char*) gosu_image->data().to_bitmap().data();

    return (const char*)&blob;
  }
```

Reply:
Generally yes, but it has to be `return blob.c_str()`.

`to_blob` is probably the only example where it annoys me that this will never free the memory. :/ I guess it's a good start though, maybe we can figure out some way to save space later.

Reply:
For strings, generally yes, but you have to `return blob.c_str()`. And you must separate the definition and initialization/assignment of the `static` variable, because otherwise the variable will only be initialized once. And in this particular case, you should use a `vector` or stick with `Gosu::Bitmap` instead of a `string`, because `to_blob` can contain null bytes, which would terminate the blob early.

```cpp
static thread_local Gosu::Bitmap bitmap;
bitmap = gosu_image->data().to_bitmap();
return reinterpret_cast<const char*>(bitmap.data());
```

`to_blob` is probably the only example where it annoys me that this will never free the memory :/ I mean, you could add a hack, like calling `Gosu_Image_to_blob(nullptr)` to free the static storage once it's been copied to Ruby ...

In an oldschool C library like Win32, usually the caller of the function will provide a memory buffer for the function to write into. Since the caller knows the width and height of the image, this would work in this case.